### PR TITLE
Move regex literal parsing logic from SwiftCompilerSources to ASTGen

### PR DESCRIPTION
**Pull request by DougGregor**
ASTGen always builds with the host Swift compiler, without requiring bootstrapping, and is enabled in more places. Move the regex literal parsing logic there so it is enabled in more host environments, and makes use of CMake's Swift support. Enable all of the regex literal tests when ASTGen is built, to ensure everything is working.

Remove the "AST" and "Parse" Swift modules from SwiftCompilerSources, because they are no longer needed.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by bnbarham — COMMENTED** [review]
Seems like maybe we should move `lib/ASTGen` to something like `swiftlib`. Or maybe this all just becomes `SwiftCompilerSources` once we've moved that across as well.

Also, woo!

**Comment by DougGregor** [comment]
@swift-ci please clean smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please clean smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/ASTGen/Sources/ASTGen/Regex.swift [hunk 1]
```diff
@@ -106,13 +113,20 @@ public func _RegexLiteralParsingFn(
...
     return false
   } catch let error as CompilerParseError {
-    var diagLoc = SourceLoc(bridged: bridgedDiagnosticBaseLoc)
-    let diagEngine = DiagnosticEngine(bridged: bridgedDiagnosticEngine)
-    if let _diagLoc = diagLoc, let errorLoc = error.location {
+    var diagLoc = bridgedDiagnosticBaseLoc
+    if let _diagLoc = diagLoc.raw, let errorLoc = error.location {
       let offset = str.utf8.distance(from: str.startIndex, to: errorLoc)
-      diagLoc = _diagLoc.advanced(by: offset)
+      diagLoc = BridgedSourceLoc(raw: _diagLoc.advanced(by: offset))
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/ASTGen/Sources/ASTGen/Regex.swift:119]: Ah, good point

File: lib/ASTGen/Sources/ASTGen/Regex.swift [hunk 2]
```diff
@@ -10,14 +10,15 @@
 //
 //===----------------------------------------------------------------------===//
 
-import _RegexParserBridging
-import AST
-import Basic
+//import AST
+//import Basic
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/ASTGen/Sources/ASTGen/Regex.swift:14]: Yes, I forgot that I'd only commented them out

**Comment by DougGregor** [comment]
> Seems like maybe we should move `lib/ASTGen` to something like `swiftlib`. Or maybe this all just becomes `SwiftCompilerSources` once we've moved that across as well.

I don't know if we want a `swiftlib` kind of thing, or a bunch of different libraries in various `lib` subdirectories. They're going to end up being intertwined with the C++ libraries.

**Comment by bnbarham** [comment]
> > Seems like maybe we should move `lib/ASTGen` to something like `swiftlib`. Or maybe this all just becomes `SwiftCompilerSources` once we've moved that across as well.
> 
> I don't know if we want a `swiftlib` kind of thing, or a bunch of different libraries in various `lib` subdirectories. They're going to end up being intertwined with the C++ libraries.

That's fine too, `lib/ASTGen` is just an odd mix now. `SwiftCompilerSources` and `lib/ASTGen` are also open-able as packages, we could do that in just `lib` as well but right now that would involve excluding all the other directories.

**Comment by finagolfin** [comment]
Thanks for doing this: I just applied this pull to the Sep. 2 trunk source snapshot and was able to build a Swift compiler natively on Android that supports regex literals again and Macros for the first time. Any plans to backport this patch to 5.9?

**Comment by DougGregor** [comment]
> Thanks for doing this: I just applied this pull to the Sep. 2 trunk source snapshot and was able to build a Swift compiler natively on Android that supports regex literals again and Macros for the first time. Any plans to backport this patch to 5.9?

Fantastic! Yes, I'm going to address review comments and back port this patch to 5.9.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by finagolfin** [comment]
I went ahead and backported this pull to 5.9, slapped it up as [a gist so you don't have to repeat that work](https://gist.github.com/finagolfin/f3d94125c06bc5db70cfb1317d5fe3ec). Haven't tried compiling it with the 5.9 branch yet, will do that next.

File: tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp [hunk 3]
```diff
@@ -111,6 +111,7 @@ void sourcekitd::initializeService(
     StringRef diagnosticDocumentationPath,
     std::function<void(sourcekitd_response_t)> postNotification) {
   INITIALIZE_LLVM();
+  initializeSwiftParseModules();
   initializeSwiftModules();
```
Comment (hamishknight) [inline tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp:115]: Would it be worth renaming `initializeSwiftModules` to something like `initializeSwiftSILModules` on the Swift side, and then defining `initializeSwiftModules` on the C side to call both that and `initializeSwiftParseModules`?

**Review by hamishknight — APPROVED** [review]
Nice!

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@finagolfin we've got Windows CI updated and macros support landed there (https://github.com/apple/swift/pull/68334), so now it's time to get this tested and merged

**Comment by DougGregor** [comment]
@swift-ci please build toolchain Windows

**Comment by finagolfin** [comment]
@DougGregor, sounds good, I'm currently cross-compiling my 5.9 version of this pull for Android. It appears that portions of #66044 will also need to be backported to 5.9, or the 5.9 version of this pull will have to be adapted for that bridging pull not being in 5.9.

**Comment by finagolfin** [comment]
Linux CI failed when testing lldb:
```
Error output:
error: error while processing module import: error: <EXPR>:7:2: error: expected a macro identifier for a pound literal expression
#/Order from <(.*)>, type: (.*), count in dozen: ([0-9]+)/#
```

**Comment by DougGregor** [comment]
@swift-ci please test Linux

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
This will need to be updated for #68408.

**Comment by DougGregor** [comment]
Dusting this off now and resolving the conflicts.

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please test

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please test

File: lib/Parse/Lexer.cpp [hunk 4]
```diff
@@ -35,11 +35,9 @@
 // Regex lexing delivered via libSwift.
 static RegexLiteralLexingFn regexLiteralLexingFn = nullptr;
 
-#if SWIFT_SWIFT_PARSER
```
Comment (bnbarham) [inline lib/Parse/Lexer.cpp]: It's not, it's just called `SWIFT_BUILD_SWIFT_SYNTAX` now. Was changed in the FetchContent PR. Do I regret that? Maybe a little.

We *could* totally get rid of it now, but I wouldn't for this PR personally.

Comment (DougGregor) [reply lib/Parse/Lexer.cpp]: ARGH. Force-pushed now that I realized this, about the same time you told me :)

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please test

File: lib/Parse/Lexer.cpp (hunk 4 above)
Comment (bnbarham) [reply lib/Parse/Lexer.cpp]: Heh, sorry for not noticing it in here in the first place!

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
The macOS test never actually got started properly. Trying again...

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please test macOS

**Comment by compnerd** [comment]
Please test with following PRs:
apple/llvm-project#7550
https://github.com/apple/swift-installer-scripts/pull/210

@swift-ci please build toolchain Windows platform

**Comment by compnerd** [comment]
Please test with following PRs:
https://github.com/apple/llvm-project/pull/7550
https://github.com/apple/swift-installer-scripts/pull/210

@swift-ci please build toolchain Windows platform

**Comment by DougGregor** [comment]
https://github.com/apple/llvm-project/pull/7550

@swift-ci please test

**Comment by finagolfin** [comment]
I have [backported this pull to Swift 5.9](https://github.com/termux/termux-packages/blob/master/packages/swift/swift-move-regex-literals.patch) and am happy to report a working Swift 5.9 package with regex literals again on Android in the Termux app. I also backported #66033 and #66044 there, as this pull relies on those changes that aren't in the 5.9 branch.

I've also backported #68408 and apple/swift-syntax#2173 and applied some changes so Swift Syntax can be cross-compiled (which I'll upstream next), so that Macros work natively on Android.

**Comment by DougGregor** [comment]
This was already done

**Pull request closed without merging**